### DISCUSSION TITLE
How to put requests from one consumer in queue (Lock feature)

### DISCUSSION THREAD

**Discussion by lavernyer** [Kong/kong#7272] [Ideas and feature requests]
I'm using kong as API gateway for frontend. How can I prevent multiple requests from frontend at the same time?

Expected options:
1. Kong put requests to the specific route from the same consumer to the queue.
2. Kong cancel requests to the specific route from the same consumer if there is active request already.

Example:
Consumer do 5 POST requests to the route '/api/order'.
1. Kong put all requests in queue and process each request in order after previous sent response.
2. Kong take first request and reject all rest requests to this specific route from that consumer while our first request is performing.

**Comment by Tieske** [comment]
Interesting use case. How should this work on a Kong cluster? Ideas?

**Comment by p0pr0ck5** [comment]
I imagine the second case would be very straightforward to do via `cluster_mutex` (or a local redis might be faster in a single DC deployment).

Queueing up requests would take a bit more logic, again, pretty easy to do with a local redis queue used and Kong consuming that queue.

Either option requires a custom plugin :)

**Comment by p0pr0ck5** [comment]
@lavernyer to be clear, neither of these options is possible today. However, it's possible to extend Kong via plugins (see https://docs.konghq.com/1.1.x/plugin-development/), so you can develop the business logic you need here and implement it via a plugin for Kong.
